src/sql/util.py

```diff
@@ -527,3 +528,33 @@ def validate_mutually_exclusive_args(arg_names, args):
             f"{pretty_print(specified_args)} are specified. "
             "You can only specify one of them."
         )
+
+
+def validate_nonidentifier_connection(args, command):
+    """
+    Raises UsageError if a connection is passed to `%sql/%%sql` through
+    object property, list, or dictionary.
+
+    Parameters
+    ----------
+    args : argparse.Namespace
+        Contains %sql command line arguments values
+    command : sql.command.SQLCommand
+        Contains sql query after variable expansion if necessary
+    """
+    command_sql_first_line = command.sql.split("\n")[0].split(" ")
+    if (
+        len(command_sql_first_line) == 1
+        and bool(command_sql_first_line[0])
+        and not command_sql_first_line[0].isidentifier()
+        and not (args.persist_replace or args.persist or args.append)
+    ):
+        try:
+            ast.parse(command_sql_first_line[0])
+            raise exceptions.UsageError(
+                f"'{command_sql_first_line[0]}' is not a valid connection identifier. "
+                "Please pass the variable's name directly, as passing "
+                "object attributes, dictionaries or lists won't work."
+            )
+        except SyntaxError:
+            pass
```

Comment:
this part is unclear, you can abstract the ast.parse into a function like `is_valid_python_code`, because having the ast.parse line and a raise  inside a try statement is confusing

Reply:
[fixed](https://github.com/ploomber/jupysql/pull/821/files#diff-653c15e6161c97f356be04e072838ed15014af275c06019a7692ab727fa44b97R543-R556) : created `is_valid_python_code` function to return `True/False` based on `ast.parse(code)` and raised the `exceptions.UsageError` inside a if statement